File: CodeSignal/SingleFunction7b.java
/*
You've created a new programming language, and now you've decided to add hashmap support to it.
Actually you are quite disappointed that in common programming languages it's impossible to add a number to all hashmap keys, or all its values. 
So you've decided to take matters into your own hands and implement your own hashmap in your new language that has the following operations:

insert x y - insert an object with key x and value y.
get x - return the value of an object with key x.
addToKey x - add x to all keys in map.
addToValue y - add y to all values in map.
To test out your new hashmap, you have a list of queries in the form of two arrays: queryTypes contains the names of the methods to be called (eg: insert, get, etc), 
and queries contains the arguments for those methods (the x and y values).

Your task is to implement this hashmap, apply the given queries, and to find the sum of all the results for get operations.
 */


long solution(String[] queryType, int[][] query) {
    // Initialize variables to keep track of the current key and value increments, and a HashMap to store key-value pairs.
    int curKeyInc = 0;
    int curValInc = 0;
    HashMap<Integer, Integer> mp = new HashMap<>();
    // Initialize a variable to store the final answer.
    long ans = 0;

    // Iterate through each query.
    for (int i = 0; i < queryType.length; i++) {
        // If the query type is "insert", add the key-value pair to the HashMap after adjusting the current increments.
        if (queryType[i].equals("insert")) {
            mp.put(query[i][0] - curKeyInc, query[i][1] - curValInc);
        }
        // If the query type is "addToValue", update the current value increment.
        else if (queryType[i].equals("addToValue")) {
            curValInc += query[i][0];
        }
        // If the query type is "addToKey", update the current key increment.
        else if (queryType[i].equals("addToKey")) {
            curKeyInc += query[i][0];
        }
        // If the query type is "get", find the corresponding value in the HashMap using the adjusted key and calculate the final result.
        else {
            ans += mp.getOrDefault(query[i][0] - curKeyInc, -1 * curValInc) + curValInc;
        }
    }

    // Return the final result.
    return ans;
}

File: CodeSignal/SingleFunction7b.py
# Define a function that adds an integer 'k' to each value in the input dictionary 'hmap'
def addToValue(hmap, k):
    for key, val in hmap.items():
        hmap[key] += k
    return hmap

# Define a function that adds an integer 'k' to each key in the input dictionary 'hmap'
def addToKey(hmap, k):
    # The keys of a dictionary are immutable, which means they cannot be modified once they are created hence we need to create a new dictionary
    newHmap = {}
    for key, val in hmap.items():
        newKey = key + k
        newHmap[newKey] = val
    return newHmap

# Define a function that processes a list of queries and returns the final answer
def solution(queryType, query):
    # Initialize a variable 'ans' to store the final result
    ans = 0
    # Initialize an empty dictionary 'hmap' to store key-value pairs
    hmap = {}
    # Initialize variables 'ck' and 'cv' to track changes in keys and values
    ck = 0
    cv = 0
    
    # Loop through the list of query types and queries
    for i in range(len(queryType)):
        # Extract the command (cmd) and the query (quer) for the current iteration
        cmd = queryType[i]
        quer = query[i]
        
        # If the command is "insert"
        if cmd == "insert":
            # Extract the key and value from the query
            key, val = quer[0], quer[1]
            # Modify the key and value based on 'ck' and 'cv'
            hmap[key - ck] = val - cv
        
        # If the command is "addToValue"
        elif cmd == "addToValue":
            # Extract the value 'k' from the query and update 'cv'
            k = quer[0]
            cv += k
        
        # If the command is "addToKey"
        elif cmd == "addToKey":
            # Extract the value 'k' from the query and update 'ck'
            k = quer[0]
            ck += k
        
        # If the command is "get"
        else:
            # Extract the value 'k' from the query and adjust 'k' based on 'ck'
            k = quer[0] - ck
            # Look up the corresponding value in 'hmap', add 'cv', and update 'ans'
            val = hmap[k] + cv
            ans = ans + val
    
    # Return the final answer
    return ans


"""
Tests passed: 30/30.
Sample tests:
15/15
Hidden tests:
15/15
Score:
500/500
"""